odin/src/main/scala/org/clulab/odin/impl/DependencyPatternCompiler.scala

```diff
@@ -146,20 +174,44 @@ class ArgumentPattern(
     val label: String,
     val pattern: DependencyPatternNode,
     val required: Boolean,
-    val size: Option[Int]
+    val quantifier: Quantifier
 ) {
   // extracts mentions and groups them according to `size`
   def extract(tok: Int, sent: Int, doc: Document, state: State): Seq[Seq[(Mention, SynPath)]] = {
     val matches = for {
       (tok, path) <- pattern.findAllIn(tok, sent, doc, state)
       m <- state.mentionsFor(sent, tok, label)
     } yield (m, path.reverse) // paths were collected in reverse
-    if (matches.isEmpty) Nil
-    else if (size.isDefined) matches.combinations(size.get).toList
-    else Seq(matches)
+    (matches, quantifier) match {
+      // no matches
+      case (Nil, _) => Nil
+      // no quantifier w/ some matches
+      case (_, NullQuantifier) => matches.combinations(1).toList
+      case (_, OptionalQuantifier) => Seq(matches)
+      case (_, ExactQuantifier(n)) => matches.combinations(n).toList
+      // Kleene star
+      case (_, KleeneStar(true)) => Seq(matches)
+      case (_, KleeneStar(false)) => Nil
+      // One or more
+      case (_, OneOrMore(true)) => Seq(matches)
+      case (_, OneOrMore(false)) => matches.combinations(1).toList
+      // ranged quantifier w/ min and max reps
+      case (_, RangedQuantifier(Some(minRep), Some(maxRep))) =>
+        val reps = (Seq(matches.size) ++ (0 to maxRep)).filter(_ <= matches.size).max
+        matches.combinations(reps).toList
```

Comment:
Consider a different way of writing this, for example, this may be more readable:
```
if (matches.size < minRep) Nil
else if (matches.size > maxRep) matches.combinations(maxRep).toList
else Seq(matches)
```